FLOE-423: Roughed in /user POST endpoint

jobara opened the pull request:
https://issues.fluidproject.org/browse/FLOE-423

jobara committed:
FLOE-423: Adding the POST endpoint for saving preferences

The actual connection to the security layer is mocked out and should be
wired up properly when the security server implements the required features.

jobara committed:
FLOE-423: Beginning of tests

The tests currently are not working, but this is a checkin of the work in progres.

duhrer committed:
FLOE-423.  Added standard grunt boilerplate to automatically run `grunt dedupe-infusion` when `npm install` is run.

jobara committed:
FLOE-423: Continued work on tests.

A bit further. The tests are running, but still fail.

jobara committed:
Merge remote-tracking branch 'the-t-in-rtf/FLOE-423' into FLOE-423

jobara committed:
FLOE-423: whitespace fixes

jobara committed:
FLOE-423: In progress commit

Attempting to fix an issue with the Grunt tasks. Also trying to refactor the url used in the test.

jobara committed:
FLOE-423: Working around deduplication issues

amb26 commented on install.js:
```diff
@@ -0,0 +1,36 @@
+            // if dedupe process exited normally
+            if (exitCode !== null) {
+                // spanws the re-install of first-discovery
+                // this is needed because the deduplication removes its version of
```
It isn't clear from this comment what problem is being solved here - can you explain further?

amb26 commented on install.js:
```diff
@@ -0,0 +1,36 @@
+// spawns the npm install
+// need to use --force to ensure that "dedupe-infusion" failures from gpii-express
+// do not prevent its installation.
+var npm = child_process.spawn("npm", ["install", "--force"], {stdio: "inherit"});
```
Looks like this repeated use of spawn already merits being factored out into a function.
It's worth adding to the comment for this function that we use spawn rather than exec because we are uninterested in the buffer from exec and it may overflow - useful guide here: http://www.hacksparrow.com/difference-between-spawn-and-exec-of-node-js-child_process.html

amb26 commented on install.js:
```diff
@@ -0,0 +1,36 @@
+var child_process = require("child_process");
+
+// spawns the npm install
+// need to use --force to ensure that "dedupe-infusion" failures from gpii-express
```
In what case does dedupe-infusion fail? What error does it report when it does?

amb26 commented on package.json:
```diff
@@ -4,7 +4,8 @@
   "main": "index.js",
   "engines": {
-    "node": ">=0.10.0"
+    "node": "0.10.36",
```
I think being forgiving on versions of node and npm (at least to the extent of 0.10.x and 2.x.x) is reasonable - even if we are deciding to be less forgiving about code dependencies

amb26 commented on package.json:
```diff
@@ -34,6 +35,9 @@
-    "jqUnit": "git+https://github.com/fluid-project/node-jqUnit.git"
+    "grunt-shell": "0.6.4",
+    "grunt-gpii": "git://github.com/GPII/grunt-gpii.git#58ae055833f437a9e91baedb9f39a22c5fdd727f",
+    "jqUnit": "git+https://github.com/fluid-project/node-jqUnit.git",
```
I imagine this should be bound to a specific version too, like the other git dependencies

amb26 commented on .jshintrc:
```diff
@@ -26,6 +26,7 @@
 
     "globals": {
-        "fluid_2_0": true
+        "fluid_2_0": true,
```
There shouldn't be any uses of the fluid_2_0 global in this codebase, and coding this here will only make us have to keep upgrading

jobara committed:
FLOE-423: Updating comments

jobara committed:
FLOE-423: Wrapping spawn to handle default options

jobara committed:
FLOE-423: modifying version dependencies

Reducing requirements for node and npm, and specifying a specific revision of jqunit

jobara committed:
FLOE-423: removing fluid_2_0 from globals

jobara commented:
@amb26 ready for more review

amb26 commented on install.js:
```diff
@@ -0,0 +1,48 @@
+// Need to use --force to ensure that "dedupe-infusion" failures from gpii-express
+// do not prevent its installation. The dedupe-infusion postinstall task in gpii-express
+// is currently failing because the grunt-gpii dependency is not installed before
+// the postinstall script is run.
```
Worth linking to the npm issue here which is responsible for the postinstall problem

amb26 commented on install.js:
```diff
@@ -0,0 +1,48 @@
+        dedupe.on("close", function (exitCode) {
+            // if dedupe process exited normally
+            if (exitCode !== null) {
+                // spanws the reinstall of first-discovery
```
typo spanws

amb26 commented on install.js:
```diff
@@ -0,0 +1,48 @@
+            if (exitCode !== null) {
+                // spanws the reinstall of first-discovery
+                // first-discovery also depends on Infusion and contains a copy of
+                // Infusion wihtin its lib directory.
```
typo wihtin

amb26 commented on install.js:
```diff
@@ -0,0 +1,48 @@
+                // Infusion wihtin its lib directory.
+                // When the deduplication process is run, this copy of Infusion
+                // is also removed. However, because first-discovery runs on the client
+                // and does not make use of "require" it cannot locate the copy of Infusion.
```
This seems like a bug in dedupe-infusion, to me. We should reimplement it so that it only deletes infusions which are directly nested in a node_modules directory

amb26 commented on tests/js/preferencesRouter-tests.js:
```diff
@@ -0,0 +1,99 @@
+require("../../src/js/firstDiscoveryServer.js");
+require("../../node_modules/gpii-express/tests/js/lib/test-helpers.js");
+require("../../node_modules/kettle");
+require("../../node_modules/kettle/lib/test/KettleTestUtils");
```
These direct "blind includes" into other packages are all unstable and shouldn't be used - they will break in future versions of npm where module layout can't be predicted.

The latter two lines should be replaced by
var kettle = fluid.registerNamespace("kettle");
kettle.loadTestingSupport();

the first require is fine since it is a file in the current module.
the require of test-helpers probably can't be replaced now but I've just filed https://issues.gpii.net/browse/GPII-1302 which you can reference here and fix up the load once it is resolved

jobara committed:
FLOE-423: Updated comments

Fixed typos and added references to issue numbers

jobara committed:
FLOE-423: Addressing inclusion issues

jobara committed:
FLOE-423: Addressed kettle includes

jobara commented:
@amb26 ready for more review

avtar committed:
FLOE-423: Makes Vagrant config using vboxsf instead of rsync

jobara committed:
FLOE-423: Setting the version of NPM to use in the VM

jobara commented:
@amb26 I've merged in @avtar's changes. This is ready for more review.

jobara commented:
@amb26 when you review this pull request, please make sure that you are testing with a clean VM (i.e. call `vagrant destroy -f` before `vagrant up`). Thanks.

jobara committed:
FLOE-423: Switched to use dedupe-infusion script

Refactoring installation to use npm scritps instead of the install.js.

jobara commented:
@amb26 I've made use of the dedupe-infusion package. Ready for more review.

jobara committed:
FLOE-423: Updated bugs URL

amb26 commented:
Thanks, @jobara - I tried this install on my (very slow) laptop and got the following error during npm install inside the VM:

==> default: TASK: [nodejs | Run application related commands] *****************************
==> default: failed: [localhost] => (item=npm run fullinstall) => {"changed": true, "cmd": ["npm", "run", "fullinstall"]
, "delta": "0:04:31.480156", "end": "2015-10-15 13:18:36.067846", "item": "npm run fullinstall", "rc": 1, "start": "2015
-10-15 13:14:04.587690", "warnings": []}
==> default: stderr: npm ERR! Linux 3.10.0-229.14.1.el7.x86_64
==> default: npm ERR! argv "node" "/bin/npm" "install"
==> default: npm ERR! node v0.10.40
==> default: npm ERR! npm  v2.14.5
==> default: npm ERR! path /home/vagrant/sync/node_modules/grunt/node_modules/rimraf/package.json.5bc290ffee1231b9b2bbab
fd228a6a75
==> default: npm ERR! code ETXTBSY
==> default: npm ERR! errno 62
==> default:
==> default: npm ERR! ETXTBSY, rename '/home/vagrant/sync/node_modules/grunt/node_modules/rimraf/package.json.5bc290ffee
1231b9b2bbabfd228a6a75'
==> default: npm ERR!
==> default: npm ERR! If you need help, you may report this error at:
==> default: npm ERR!     https://github.com/npm/npm/issues
==> default: npm ERR! Linux 3.10.0-229.14.1.el7.x86_64
==> default: npm ERR! argv "node" "/bin/npm" "install"
==> default: npm ERR! node v0.10.40
==> default: npm ERR! npm  v2.14.5
==> default: npm ERR! path npm-debug.log.9858cac0aaf4c8170df616b1a1398c24
==> default: npm ERR! code ETXTBSY
==> default: npm ERR! errno 62
==> default:
==> default: npm ERR! ETXTBSY, rename 'npm-debug.log.9858cac0aaf4c8170df616b1a1398c24'
==> default: npm ERR!
==> default: npm ERR! If you need help, you may report this error at:
==> default: npm ERR!     https://github.com/npm/npm/issues
==> default:
==> default: npm ERR! Please include the following file with any support request:
==> default: npm ERR!     /home/vagrant/sync/npm-debug.log
==> default:
==> default: npm ERR! Linux 3.10.0-229.14.1.el7.x86_64
==> default: npm ERR! argv "node" "/bin/npm" "run" "fullinstall"
==> default: npm ERR! node v0.10.40
==> default: npm ERR! npm  v2.14.5
==> default: npm ERR! code ELIFECYCLE
==> default: npm ERR! first-discovery-server@0.1.0-SNAPSHOT fullinstall: `npm install && npm run dedupe-infusion`
==> default: npm ERR! Exit status 62

I'll try again when I get to my desktop...

amb26 commented:
Looks like an uncharacterised and unfixed npm bug: https://github.com/npm/npm/issues/8134 - I love working with nerds :)

jobara committed:
FLUID-423: Updated readme

amb26 commented on index.js:
```diff
@@ -10,4 +10,8 @@ https://github.com/fluid-project/first-discovery-server/raw/master/LICENSE.txt
+var gpii = fluid.registerNamespace("gpii");
 require("./src/js/firstDiscoveryServer.js");
+
+gpii.firstDiscovery.server();
```
Now you've moved the startup into this driver file, I think this is a good place for making the configuration for this server to be sensitive to the vagrant configuration for the port - which is the ultimate authoritative upstream value. @avtar  has arranged for it to be read into an environment variable. A reasonable approach would be to send a distributeOptions block into the arguments of this function call which forwards the port number to the correct server. You should also consult with @duhrer about whether configuration for gpii-express can be improved so that we don't need to write the port number in two places. What happens if we leave out the "baseUrl"?

the-t-in-rtf replied:
I have done a lot of work in recent text fixtures to propagate the port into sensible base URLs, it's certainly possible to move some of this into the defaults for gpii.express.

amb26 replied:
@the-t-in-rtf What would it look like if done so?

the-t-in-rtf replied:
Basically an expander that calls fluid.stringTemplate and updates the baseUrl value with the port, as seen here:

https://github.com/the-t-in-rtf/gpii-express-user/blob/master/tests/js/test-environment.js

jobara committed:
FLUID-423: Use port specified by ENV

jobara commented:
@amb26 ready for more review.

_**NOTE**: if you change the port number in vars.yml after you have already created a vm. You'll need to run `vagrant reload` and possibly `vagrant provision`._

amb26 commented on Vagrantfile:
```diff
@@ -1,33 +1,70 @@
+# environment variables. Otherwise if 'nodejs_app_tcp_port' is defined
+# in vars.yml then use that port. Failing that use defaults provided
+# in this file.
+host_tcp_port = ENV["VM_HOST_TCP_PORT"] || ansible_vars["nodejs_app_tcp_port"] || 8080
```
We need to actually use the port that the user suggested! There needs to be a single place where the user changes the setting, and all places should respond to that setting - rather than the current situation where there are apparently 4 uncoordinated environment variables. We should also default to a port other than 8080.

jobara replied:
@avtar thoughts?

jobara replied:
@amb26  I changed the default port to 8088. I'm not sure I understand what the issue is with the environment variables though. Although I may be misunderstanding how they are all used. For dev use you'd just change `nodes_app_tcp_port` in vars.yml. If you weren't using the VM you could set the `FIRST_DISCOVERY_SERVER_TCP_PORT` environment variable instead.

amb26 replied:
@jobara - the issue is one of usability. If the user "just changes nodejs_app_tcp_port in vars.yml" they should expect it to actually work - that is, the application should actually use that port. Whereas in fact what happens if they just change that entry, the application will fail to start up since it is in fact using a different port.
At the very least, there should be clear instructions on EVERYTHING the user needs to do to make a working instance of the app come up and actually run, inside the vagrant box on a different port - and, preferably, those instructions should consist of making only one change to a configuration file.

avtar replied:
@amb26, @jobara the `VM_*_TCP_PORT` environment variables for overriding ports were provided so someone could launch the VM and select a different port without having to edit `vars.yml` or changing the default in this `Vagrantfile`. However I think they can be removed now since port conflicts are prevented. If someone is using 8088 with another process on their machine it they can still start this VM and just use the new port that Vagrant logs in its output at startup.

amb26 commented on tests/js/preferencesRouter-tests.js:
```diff
@@ -0,0 +1,103 @@
+// Once https://issues.gpii.net/browse/GPII-1302 has been addressed
+// the following blind include should be replaced by a more robust
+// inclusion mechanism.
+require("../../node_modules/gpii-express/tests/js/lib/test-helpers.js");
```
My understanding of the node require system suggests that this include can already be replaced by 

require("gpii-express/tests/js/lib/test-helpers.js")

amb26 commented on tests/js/preferencesRouter-tests.js:
```diff
@@ -0,0 +1,103 @@
+
+"use strict";
+
+var fluid = fluid || require("infusion");
```
var fluid = require("infusion") is the standard include for node-aware files

amb26 commented on src/js/preferencesRouter.js:
```diff
@@ -0,0 +1,46 @@
+
+"use strict";
+
+var fluid = fluid || require("infusion");
```
See below

amb26 commented on src/js/firstDiscoveryServer.js:
```diff
@@ -11,9 +11,9 @@ https://github.com/fluid-project/first-discovery-server/raw/master/LICENSE.txt
 "use strict";
 
 var fluid = fluid || require("infusion");
```
Fix require

jobara committed:
FLOE-423: Fixed port in tests

jobara committed:
FLOE-423: Updated require statements

jobara committed:
FLOE-423: Changed default port to 8088

jobara committed:
FLOE-423: Updating port forwarding settings

jobara commented:
@amb26 I've modified the port configuration based on your and @avtar's recommendations. This pull request is ready for more review.

jobara committed:
Merge branch 'master' into FLOE-423

amb26 commented on README.md:
```diff
@@ -13,6 +24,8 @@ Vagrant and Ansible configuration has been provided to make the setup of a devel
-Once you've cloned the repository onto your local system you'll only need to run `vagrant up` to create the vm. The server will be accessible on your host machine at `http://localhost:8080`.
+Once you've cloned the repository onto your local system you'll only need to run `vagrant up` to create the vm. By default, the server will be accessible on your host machine at `http://localhost:8080`. (e.g. `http://localhost:8080/demos`).
+
+**_NOTE_: If you changed the port option, `nodejs_app_tcp_port`, in the vars.yml file or 8080 is already in use on your host machine, the actual URL may be different.**
```
It looks like the default port is now actually 8088

amb26 commented on src/js/firstDiscoveryServer.js:
```diff
@@ -21,10 +21,15 @@ var fdSrcDir = path.resolve(__dirname, "../../node_modules/first-discovery/src")
 
 fluid.defaults("gpii.firstDiscovery.server", {
     gradeNames: ["gpii.express"],
+    port: 8080,
```
This default doesn't agree with the one in vars.yml and there still seems to be no attempt to keep these two port numbers in step. Recall that the mission is to provide a "one stop shop" - a single place in a single configuration file that the user needs to change in order to get a fully functioning server inside the container, running on a port of their choosing

jobara replied:
I changed the default to 8088. The "one stop shop" is taken care of in the index.js file. (see: https://github.com/jobara/first-discovery-server/blob/FLOE-423/index.js#L17-L19)

jobara committed:
FLOE-423: Updated default port number in readme

jobara committed:
FLOE-423: Updated default port number to 8088

jobara commented:
@amb26 ready for more review.
